src/lib/cht-session.ts

```diff
@@ -0,0 +1,102 @@
+const axios = require('axios'); // require is needed for rewire
+
+import { AuthenticationInfo } from '../config';
+import { AxiosHeaders } from 'axios';
+import { RemotePlace } from './cht-api';
+
+const COUCH_AUTH_COOKIE_NAME = 'AuthSession=';
+const ADMIN_FACILITY_ID = '*';
+
+export default class ChtSession {
+  public readonly authInfo: AuthenticationInfo;
+  public readonly sessionToken: string;
+  public readonly username: string;
+  public readonly facilityId: string;
+
+  private constructor(authInfo: AuthenticationInfo, sessionToken: string, username: string, facilityId: string) {
+    this.authInfo = authInfo;
+    this.sessionToken = sessionToken;
+    this.username = username;
+    this.facilityId = facilityId;
+
+    if (!this.sessionToken || !this.authInfo.domain || !this.username || !this.facilityId) {
+      throw new Error('invalid CHT session information');
+    }
+  }
+
+  public static async create(authInfo: AuthenticationInfo, username : string, password: string): Promise<ChtSession> {
+    const sessionToken = await createSessionToken(authInfo, username, password);
+
+    if (!sessionToken) {
+      throw new Error(`failed to obtain token for ${username} at ${authInfo.domain}`);
+    }
+    
+    const userDetails = await fetchUserDetails(authInfo, username, sessionToken);
+    const facilityId = userDetails.isAdmin ? ADMIN_FACILITY_ID : userDetails.facilityId;
+    if (!facilityId) {
+      throw Error(`User ${username} does not have a facility_id connected to their user doc`);
+    }
+    
+    return new ChtSession(authInfo, sessionToken, username, facilityId);
+  }
+
+  public static createFromDataString(data: string): ChtSession {
+    const parsed:any = JSON.parse(data);
+    return new ChtSession(parsed.authInfo, parsed.sessionToken, parsed.username, parsed.facilityId);
+  }
+
+  clone(): ChtSession {
+    return new ChtSession(this.authInfo, this.sessionToken, this.username, this.facilityId);
+  }
+
+  isPlaceAuthorized(remotePlace: RemotePlace): boolean {
+    return !!this.facilityId &&
+      (
+        this.facilityId === ADMIN_FACILITY_ID 
+        || remotePlace?.lineage?.includes(this.facilityId)
+        || remotePlace?.id === this.facilityId
+      );
+  }
+}
+
+async function createSessionToken(authInfo: AuthenticationInfo, username: string, password: string): Promise<string> {
```

Comment:
Would it be better to refactor this function,  `fetchUserDetails` and `createUrl` as a private static method within `ChtSession` as there are used only in that function?

Reply:
I'm not sure if I prefer having the function private because it is in the module scope or because it is explicitly private in the class scope... but I'm happy to update as you have suggested.